crates/weblsp/src/main.rs

```diff
@@ -49,34 +49,25 @@ fn main_loop(
     init_params: serde_json::Value,
     mut css_language_service: csslsrs::service::LanguageService,
 ) -> Result<ExitCode, Box<dyn Error + Sync + Send>> {
-    let mut awaiting_exit = false;
     let _init_params: InitializeParams = serde_json::from_value(init_params).unwrap();
 
     for msg in &connection.receiver {
-        // TODO: Handle trace levels and notifications instead of just printing them to stderr.
-        eprintln!("new msg: {:?}", msg);
-
-        // If we're waiting for an exit notification, any message other than it is an error, and will cause the server to exit with a failure exit code.
-        // As such, we can handle this outside of the match statement.
-        if awaiting_exit {
-            if let Message::Notification(not) = &msg {
-                if not.method == lsp_types::notification::Exit::METHOD {
-                    return Ok(ExitCode::SUCCESS);
-                }
-            }
-            eprintln!("Shutting down without receiving `Exit` notification.");
-            return Ok(ExitCode::FAILURE);
-        }
-
         // Handle the rest of the messages.
         match msg {
             Message::Request(req) => {
-                let request =
-                    requests::handle_request(req, &mut css_language_service, &connection)?;
+                match connection.handle_shutdown(&req) {
```

Comment:
didn't know this existed when I made the initial shutdown support, turns out this does the same as what I did manually. We still need a global state eventually